cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java

```diff
@@ -51,23 +75,54 @@ public void testManyMacrosInAppSpec() {
+                                          Collections.<String, StreamSpecification>emptyMap(),
+                                          Collections.<String, String>emptyMap(),
+                                          Collections.<String, DatasetCreationSpec>emptyMap(),
+                                          Collections.<String, FlowSpecification>emptyMap(),
+                                          Collections.<String, MapReduceSpecification>emptyMap(),
+                                          Collections.<String, SparkSpecification>emptyMap(),
+                                          Collections.<String, WorkflowSpecification>emptyMap(),
+                                          Collections.<String, ServiceSpecification>emptyMap(),
+                                          Collections.<String, ScheduleSpecification>emptyMap(),
+                                          Collections.<String, ScheduleCreationSpec>emptyMap(),
+                                          Collections.<String, WorkerSpecification>emptyMap(),
+                                          Collections.<String, Plugin>emptyMap());
     cfg.setApplicationSpecification(appSpec);
-    Assert.assertEquals(appSpec.getConfiguration(), cfg.getApplicationSpecification().getConfiguration());
+    Assert.assertEquals(appSpec.getConfiguration(),
+                        cfg.getApplicationSpecification().getConfiguration());
+  }
+
+  @Test
+  public void testGetPluginsWithMacrosMoreThan20() {
+    Configuration hConf = new Configuration();
+    MapReduceContextConfig cfg = new MapReduceContextConfig(hConf);
+    Map<String, Plugin> mockPlugins = new HashMap<String, Plugin>();
+    ArtifactId artifactId = new ArtifactId("plugins", new ArtifactVersion("1.0.0"), ArtifactScope.SYSTEM);
+    Map<String, String> properties = new HashMap<String, String>();
+    properties.put("path",
+                   "${input.directory}/${a}${b}${c}${d}${e}${f}${g}${h}${i}${j}"
+                     + "${k}${l}${m}${n}${o}${p}${q}${r}${s}${t}${u}${v}${w}${x}${y}${z}.txt");
+    hConf.set("input.directory", "/dummy/path");
+    String[] alphabetsArr = {"a", "b", "c", "d", "e", "f", "g", "h",
+      "i", "j", "k", "l", "m", "n", "o", "p", "q",
+      "r", "s", "t", "u", "v", "w", "x", "y", "z"};
+    for (String alphabet : alphabetsArr) {
+      hConf.set(alphabet, alphabet);
+    }
+    Set<ArtifactId> parents = new LinkedHashSet<>();
+    Plugin filePlugin1 = new Plugin(parents, artifactId,
+                                    new PluginClass("type", "name", "desc", "clsname", "cfgfield",
+                                                    Collections.<String, PluginPropertyField>emptyMap()),
+                                    PluginProperties.builder().addAll(properties).build());
+
+    mockPlugins.put("File1", filePlugin1);
+    hConf.set(HCONF_ATTR_PLUGINS, GSON.toJson(mockPlugins, PLUGIN_MAP_TYPE));
```

Comment:
don't think PLUGIN_MAP_TYPE is needed. you can just call GSON.toJson(mockPlugins)